activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb

```diff
@@ -294,15 +294,33 @@ def initialize(pool, frequency)
           @frequency = frequency
         end
 
+        @@mutex = Mutex.new
+        @@pools = {}
+
+        def self.register_pool(pool, frequency) # :nodoc:
+          @@mutex.synchronize do
+            if @@pools.key?(frequency)
+              @@pools[frequency] << pool
+            else
+              @@pools[frequency] = [pool]
+              Thread.new(frequency) do |t|
+                loop do
+                  sleep t
+                  @@mutex.synchronize do
```

Comment:
Why the mutex here? It seems to be that we any shared resource?

Reply:
We're [iterating over `@@pools[frequency]`](https://github.com/rails/rails/pull/36296/files#diff-7735c67f539e7d1e2908a4d5d8909c24R310), which is also modified [here](https://github.com/rails/rails/pull/36296/files#diff-7735c67f539e7d1e2908a4d5d8909c24R303)